src/onnxruntime.cc

```diff
@@ -1817,7 +1856,7 @@ ModelInstanceState::SetInputTensors(
     else {
       batchn_shape =
           std::vector<int64_t>(input_shape, input_shape + input_dims_count);
-      if (max_batch_size != 0) {
+      if (max_batch_size != 0 && !is_initializer_tensor) {
```

Comment:
You probably don't need to condition on this explicitly, if the model config is successfully validated, then the model shouldn't allow dynamic batching since it is the current limitation of exposing overridable initializer as input. You may enforce it explicitly in validate function that when an input is marked optional, `model_state_->MaxBatchSize()` must be 0.

Reply:
Then the new API and surrounding work in the core repo may be removed.